### DISCUSSION TITLE
how to locate the Python interpreter used by QGIS?

### DISCUSSION THREAD

**Discussion by sunt05** [UMEP-dev/UMEP#408] [Python]
Occasionally, QGIS users may need to update the python packages for own purposes – locating the Python interpreter actually used by QGIS is a prerequisite.

Here we share a python snippet used in our code for this, which hopefully may provide a general workaround for all platforms.

1.  launch QGIS and navigate to its python console;
![image](https://user-images.githubusercontent.com/1802656/165924150-3bd49ce9-ed6b-4c74-a567-70eda1a45048.png)

2. save the the following code using the console's editor:
```python
import sys, subprocess, os
from pathlib import Path
import platform
from packaging import version

def locate_py():

    # get Python version
    str_ver_qgis = sys.version.split(' ')[0]
    
    try:
        # non-Linux
        path_py = os.environ["PYTHONHOME"]
    except Exception:
        # Linux
        path_py = sys.executable

    # convert to Path for eaiser processing
    path_py = Path(path_py)

    # pre-defined paths for python executable
    dict_pybin = {
        "Darwin": path_py / "bin" / "python3",
        "Windows": path_py / ("../../bin/pythonw.exe" if version.parse(str_ver_qgis) >= version.parse("3.9.1") else "pythonw.exe"),
        "Linux": path_py,
    }

    # python executable
    path_pybin = dict_pybin[platform.system()]

    if path_pybin.exists():
        return path_pybin
    else:
        raise RuntimeError("UMEP cannot locate the Python interpreter used by QGIS!")

```
and press the "run" button (i.e., green triangle)

3. back to the console prompt and run this code `locate_py()`,
![image](https://user-images.githubusercontent.com/1802656/165924606-8906a3cc-ccc9-4be8-94a1-7474d932bdbb.png)

 which should give something like this (e.g. on macOS):

`PosixPath('/Applications/QGIS.app/Contents/MacOS/bin/python3')`

This is the actual path to python interpreter used by your QGIS.

**Comment by robertschweizer** [comment]
Thanks for this guideline!

It might need a slight update: Using QGIS 3.34.0 from conda-forge on an ARM64 Mac, the python executable is in fact next to `sys.executable`, so in `Path(sys.executable).with_name("python3")`.
